lib/temporal/workflow/executor.rb

```diff
@@ -33,7 +33,7 @@ def initialize(workflow_class, history, task_metadata, config, track_stack_trace
 
       def run
         dispatcher.register_handler(
-          History::EventTarget.workflow,
+          History::EventTarget.start_workflow,
```

Comment:
Curious what's going on here.  I ask because it's pretty central code.

Reply:
The non-determinism checking that happens in `discard_command` only considers the "target" of a command/event and only includes it and the event ID in its error messages. Prior to this change, any operation affecting a workflow had the same target and therefore the same target in any error messages. I've specialized these targets to individual ones for starting, completing, failing and continuing as new. This comes with two major benefits.

First, the non-determinism checking now applies between these different cases. This means a history that succeeded but on replay tries to fail or continue as new will now be considered non-deterministic while previously this would have been acceptable

Second, the error message will now differentiate between these cases rather than a generic "workflow" error that may be non-obvious to the user. E.g., this error message:
> Unexpected command. The replaying code is issuing: (23, workflow) but the history of previous executions recorded: (23, activity)

becomes something like,

> Unexpected command. The replaying code is issuing: (23, fail_workflow) but the history of previous executions recorded: (23, activity)

On top of that, we now call `discard_command` in the handlers for these workflow ending events, which means we do non-determinism checking on workflow end. This is important in catching cases where a workflow history ends but on replay the code keeps running.

This .particular change is simply a consequence of that specialization of event targets. I could have left it in place, but felt it would be more confusing to see a `workflow` target only for starting workflows, while the other cases all have their own targets.